Styhead support

zboszor opened the pull request:
This is a rebase of #225 over current master.

In the meantime, the styhead branches were created in upstream Yocto and [styhead already diverged from current master](https://lists.openembedded.org/g/openembedded-core/message/205173) leading to Yocto 5.2 "Walnascar".

I added the necessary CI change to use the styhead branch from poky. Not sure if this also uses (the currently non-existing) styhead branch from meta-mono, or as intended, its master branch.

ajlennon commented:
Building was blocked again. Have restarted the container

ajlennon commented:
It seems to have errored

https://github.com/DynamicDevices/meta-mono/actions/runs/11159243181/job/31017239139?pr=231

zboszor commented:
> It seems to have errored
> 
> https://github.com/DynamicDevices/meta-mono/actions/runs/11159243181/job/31017239139?pr=231

I rebased over your `README.md` changes and added a new commit to fix all the remaining `WORKDIR` instances.

ajlennon commented:
More build errors - 

https://github.com/DynamicDevices/meta-mono/actions/runs/11179769792/job/31080098765

zboszor commented:
> More build errors -
> 
> https://github.com/DynamicDevices/meta-mono/actions/runs/11179769792/job/31080098765

These are already fixed. Let's see if the last commit goes through.

zboszor commented:
The errors are:
- gtk-sharp fails due to a newer stricter GCC 
- mono package_qa failure, apparently the TMPDIR references are now hard errors

I will try to fix these the next week.

zboszor commented:
Either we can ignore the errors with a series of `INSANE_SKIP:<subpackage> = "buildpaths"` or implement #161

ajlennon commented:
>Either we can ignore the errors with a series of INSANE_SKIP:<subpackage> = "buildpaths" or implement https://github.com/DynamicDevices/meta-mono/issues/161

I'm in your hands here @zboszor. On the one hand the INSANE_SKIP seems like a bit of a hack (although I recognise this just means it would be working the way it always has worked). On the other hand the other option feels like it might be a fair bit of work.

I'll go with whatever you suggest?

zboszor commented:
> I'll go with whatever you suggest?

I am already working on the second option using `/pathmap:`, I just had to figure out why I get
```
error CS8101: The pathmap option was incorrectly formatted
```
TL;DR: `csc` and `mcs` compilers do not like empty on the right side of `=`

Then I was getting a differerent compiler error when building `class/lib/build-linux/mscorlib.dll` related to incorrectly paired `{}`, which I patched.

Now I am getting `error CS0227: Unsafe code may only appear if compiling with /unsafe` errors, so `/unsafe` must be passed to the C# compiler to allow code with `unsafe` keyword (such as the sources for `mscorlib`) to be built.

I am getting somewhere with very small changes for now.

zboszor commented:
I had to take a turn in the approach, which indeed took more time to find and fix everything.

It turned out that passing `LOCAL_MCS_FLAGS = /pathmap:...` externally didn't get anywhere because `Makefile`s in the source have been overriding it.

Instead, I had to patch the offending `Makefile`s, patch `monodoc.dll.config` in `do_install` and delete one PDB file to avoid adding `INSANE_SKIP:mono-xbuild = "buildpaths"`

It works on my side with x86-64 host and target, let's see what CI has to say.

ajlennon commented:
Fingers crossed!

zboszor commented:
There are no more failures. Especially this is what we were waiting for:
```
NOTE: recipe mono-6.12.0.182-r0: task do_package_qa: Succeeded
```
But the CI seems to be stuck again.

ajlennon commented:
Great news! I am keeping an eye on the builds here too. I assume the rest will all go through and then I'll merge in. Thanks for your efforts as ever @zboszor !

ajlennon commented:
NB. Once that goes into master I'll also make a new styhead branch based on that.

zboszor commented:
As a side note, this last patch may be backported to the scarthgap because this way the mono build becomes reproducible.

Probably some of the other changes, except the ones with the WORKDIR -> UNPACKDIR transition that is styhead+ only.

zboszor commented:
Wow. GCC 14 does not like 32-bit arm.

ajlennon commented:
>As a side note, this last patch may be backported to the scarthgap branch because this way the mono build becomes reproducible.

That's fantastic. Love it. I have limited time with all the stuff going on here but would love to get on top of fully reproducible Yocto builds for our devices

>Don't we want to clean them up eventually?

I am sure we do. All I can say is that we're all giving our time here and whilst I am relying on the CI to flag up big problems I think my current thinking is that things will inevitably break somewhere and if it's a problem somebody will flag it and then we can figure out what's happened and ideally add a new test to the CI. I'm quite keen on adding new CI tests as we go along to try to improve regression testing.

>Wow. GCC 14 does not like 32-bit arm.

Yikes. Well I guess the positive is at least we're seeing this now!

zboszor commented:
CI is green now.

ajlennon commented:
So I created a discussion about dropping support for ARM32. What do we think?

zboszor commented:
> So I created a discussion about dropping support for ARM32. What do we think?

I am not very familiar with ARM32 but it is possible that other subarchitectures have no problem, it's just the `cortexa15t2hf-neon-poky-linux-gnueabi` BSP has problems.

FWIW, the ARM32 CI test can be easily revived when the GCC 14 build is fixed in OE-Core.

ajlennon commented:
Merged in then. Thanks !
